releases-implementation/README.md

```diff
@@ -0,0 +1,415 @@
+- When the app version bump is tested, it is not necessarily tested on top of the latest release, but actually on top of the latest release plus multiple merged, but unreleased changes in the cluster-$provider app.
+- After the app version bump is merged, it may be released right away, but it may also take some time until it is released, depending on the current state of the cluster-$provider app repo and what unreleased changes it has.
+- Here the team that owns the app is dependant on the provider-integration team and/or on the provider-independent team, because they are needed for the PR reviews and the new release. This practically means that the team that owns the app does not fully own the delivery of the app. More cross-team reviews, more cross-team communication required, more friction.
+
+Now let’s see how the above scenario would look like with the releases repository.
+- The team that owns the app opens a new PR in the releases repo where they create one or more new releases.
+- In the simplest case, app version bump is the only change in the new releases, so it is tested and delivered fully independently.
+- In case there is an existing/draft PR in the release repo where new releases are being added with new versions of other apps, that existing PR can be updated.
+	- If there is a need to have the new app version tested and delivered ASAP, e.g. in case of CVEs, the existing/draft releases repo PR does not have to be updated, and another PR can be created specifically for the new app version (after this the previously created PR can be rebased/updated).
+- The delivery of the new app versions is not entangled with the delivery of cluster-$provider app Helm chart changes.
+- The team that owns the app can create the new release on their own and they do not need to ask other teams for the review/approval, since the unit tests in the releases repository verify that the newly created releases respects the defined release rules, e.g. new app patch delivered to all providers, new app minor must be delivered in a new minor release, etc.
+
+Changes to the current e2e tests would be minimal initially.
+- Testing of cluster-$provider app would require one addition, which would be:
+	- Looking up the latest release and setting the release label on the App CR.
+	- Like today, cluster-test-suites would use custom version of cluster-$provider app from the branch it is testing (in which case the mutating webhook would not set the cluster-$provider app version based on the Release CR).
```

Comment:
How would the webhook know not to trigger?

Reply:
A mutating webhook sets `.spec.version` only if it is not set already, which is useful and even needed for testing. [Here](https://github.com/giantswarm/app-admission-controller/blob/capi-releases-poc/pkg/app/mutate_cluster_app.go) is the code.

In addition to the mutating webhook, we would probably need a validating webhook, to check cluster App `.spec.version` and the version of the `cluster-$provider` app in the Release (they must be either the same, or `.spec.version` is a dev build on top of the version that we have in the Release, to allow for overriding during testing).

Reply:
> A mutating webhook sets .spec.version only if it is not set already, which is useful and even needed for testing.

Gotta fix this so it also updates `.spec.version` when release label is updated, except when we want to override cluster-aws version for testing (which we can signal with a Helm value, which should be fine, as we already have some testing-related ephemeral values).